### PR TITLE
Reference OTA Provider app

### LINKED ISSUE #2864: Implement BDX synchronous sender role over CRMP
#### Problem
CHIP defines means for performing a bulk-data transfer (BDX) over the CHIP reliable messaging protocol (CRMP). The BDX synchronous sender role over CRMP logic is not currently implemented. 

 #### Proposed Solution
Implement BDX synchronous sender role over CRMP

### LINKED ISSUE #7283: Simple OTA Server Example/Test Project
#### Problem
In order to test OTA Server Cluster, there needs to be an example project with the following features:
* respond to all OTA Server Cluster commands
* user can set available image(s) to serve (starting with CLI filepath argument)
* serve image over BDX

Additional features to consider (or future work):
* implement a database to store available mock images
* other download methods
* track OTA statistics
* AnnounceOtaServer command
* user consent
* fine grain control over download + image application timing, "Busy" status, etc

 #### Proposed Solution
Implement the above in a new example app `ota-server`

### PR DESCRIPTION
**Pull request by holbrookt**
#### Problem
* Provide a new application for implementing OTA Provider (server).
* Fixes #7283

#### Change overview
* add `examples/ota-provider-app/`
* fix `OTAProviderDelegate` to use `CommandHandler` for sending responses

#### Testing
Tested using chip-tool:
- verified responses to all OTA Provider commands
    - some are default
    - QueryImageResponse contains bdx URI and Available/NotAvailable depending on whether or not the user provided a filepath

Testing against local branch with `ota-requestor-app`:
- verified BDX transfer can run to completion, serving local file

### CODE REVIEW

File: examples/ota-provider-app/linux/main.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    ChipLogDetail(SoftwareUpdate, "using OTA file: %s", gOtaFilepath ? gOtaFilepath : "(none)");
+
+    if (gOtaFilepath != nullptr)
+    {
+        otaProvider.SetOTAFilePath(gOtaFilepath, strlen(gOtaFilepath));
+        bdxServer.SetFilepath(gOtaFilepath, strlen(gOtaFilepath));
+    }
+
+    // TODO: is there any way to not hardcode the endpoint id?
```
Comment (holbrookt) [inline examples/ota-provider-app/linux/main.cpp:134]: Open question: is there an API that can return the endpoint ID that OTA Provider is on?

File: examples/ota-provider-app/linux/main.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+bool HandleOptions(const char * aProgram, OptionSet * aOptions, int aIdentifier, const char * aName, const char * aValue)
+{
+    bool retval = true;
+    std::ifstream ifs;
+
+    switch (aIdentifier)
+    {
+
+    case kOptionFilepath:
+        ifs.open(aValue, std::ifstream::in);
```
Comment (andy31415) [inline examples/ota-provider-app/linux/main.cpp:61]: nit: would using `access` from `<unistd.h>` be more readable? We seem to want to do a R_OK test here.

File: examples/ota-provider-app/linux/main.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    if (!chip::ArgParser::ParseArgs(argv[0], argc, argv, allOptions))
+    {
+        return 1;
+    }
+
+    DeviceLayer::ConfigurationMgr().LogDeviceConfig();
+    InitServer();
+
+    exchangeMgr = chip::ExchangeManager();
+    err         = exchangeMgr->RegisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id, &bdxServer);
```
Comment (andy31415) [inline examples/ota-provider-app/linux/main.cpp:124]: do we need to check error result here?

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+#include <string.h>
+
+using namespace chip::bdx;
+
+void BdxOtaSender::SetFilepath(const char * path, size_t length)
+{
+    if (path != nullptr)
+    {
+        strcpy(mFilepath, path);
```
Comment (andy31415) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:35]: should the length be used?

`ChipMemString.h` can be used to ensure zero terminator and no memory overflow on mFilePath.

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        TransferSession::TransferAcceptData acceptData;
+        acceptData.ControlMode  = TransferControlFlags::kReceiverDrive; // OTA must use receiver drive
+        acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+        acceptData.StartOffset  = mTransfer.GetStartOffset();
+        acceptData.Length       = mTransfer.GetTransferLength();
+        VerifyOrReturn(mTransfer.AcceptTransfer(acceptData) == CHIP_NO_ERROR,
+                       ChipLogError(BDX, "%s: %s", __FUNCTION__, chip::ErrorStr(err)));
+        break;
+    }
+    break;
```
Comment (andy31415) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:77]: These breaks seem off - probably should be deleted.

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        uint16_t bytesToRead = blockSize;
+
+        // TODO: This should be a utility function in TransferSession
+        if (mTransfer.GetTransferLength() > 0 && mNumBytesSent + blockSize > mTransfer.GetTransferLength())
+        {
+            // cast should be safe because of condition above
+            bytesToRead = static_cast<uint16_t>(mTransfer.GetTransferLength() - mNumBytesSent);
+        }
+
+        chip::System::PacketBufferHandle blockBuf = chip::System::PacketBufferHandle::New(bytesToRead);
```
Comment (andy31415) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:91]: check for null of blockBuf here.

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+        std::ifstream otaFile(mFilepath, std::ifstream::in);
```
Comment (andy31415) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:93]: do we need a `.good()` check somewhere here?

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    for (size_t i = 0; i < bufSize; ++i)
+    {
+        buf[i] = GetRandU8();
+    }
+}
+
+bool GenerateBdxUri(const Span<char> & fileDesignator, Span<char> outUri, size_t availableSize)
+{
+    const size_t kBdxPrefixLen          = 7;
+    const char bdxPrefix[kBdxPrefixLen] = "bdx://";
```
Comment (andy31415) [inline examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:59]: does ` const char bdxPrefix[] = "bdx://"` work too not need the prefix len constant?

**Review by andy31415 — APPROVED** [review]

File: examples/ota-provider-app/linux/main.cpp (hunk 2 above)
Comment (holbrookt) [reply examples/ota-provider-app/linux/main.cpp:61]: yes, this seems like what I'm looking for here

File: examples/ota-provider-app/linux/main.cpp (hunk 3 above)
Comment (holbrookt) [reply examples/ota-provider-app/linux/main.cpp:124]: wouldn't hurt

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 4 above)
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:35]: I've been looking for something like `ChipMemString.h`! I'll see how I can integrate it

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 5 above)
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:77]: Good catch, not sure where those came from.

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 6 above)
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:91]: Done

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 7 above)
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:93]: Sure, that's best

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp (hunk 8 above)
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp:59]: yes you're correct.

File: examples/ota-provider-app/linux/main.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+constexpr uint32_t kBdxPollFreqMs   = 500;
+
+bool HandleOptions(const char * aProgram, OptionSet * aOptions, int aIdentifier, const char * aName, const char * aValue)
+{
+    bool retval = true;
+
+    switch (aIdentifier)
+    {
+    case kOptionFilepath:
+        if (0 == access(aValue, R_OK))
```
Comment (jelderton) [inline examples/ota-provider-app/linux/main.cpp:60]: I thought access() returns 0 on success.  Am I missing something here?

Comment (holbrookt) [reply examples/ota-provider-app/linux/main.cpp:60]: Thanks for catching this - you're correct.

**Review by mspang — APPROVED** [review]

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+#include <core/CHIPError.h>
+#include <messaging/ExchangeContext.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <support/BitFlags.h>
+#include <support/CHIPMemString.h>
+
+#include <fstream>
+
+using namespace chip::bdx;
```
[the remark this answers is not in the document]
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:28]: Yes, I'll change this

File: examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+{
+    for (size_t i = 0; i < bufSize; ++i)
+    {
+        buf[i] = GetRandU8();
+    }
+}
+
+bool GenerateBdxUri(const Span<char> & fileDesignator, Span<char> outUri, size_t availableSize)
+{
+    const char bdxPrefix[] = "bdx://";
```
[the remark this answers is not in the document]
Comment (holbrookt) [reply examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp]: done

**Comment by holbrookt** [comment]
@andy31415 @mspang @jelderton Can one or you label this with "technical review complete"?

**Review by woody-apple — APPROVED** [review]

**Pull request merged**